SOLR-16392: Refactor and update v2 ADDREPLICAPROP API

gerlowskija opened the pull request:
# Description

The Solr community is engaged in an ongoing effort to make Solr's v2 API more REST-ful and intuitive for users so that it can eventually serve as a viable v1 replacement.  This PR in particular focuses on Solr's "add-replica-prop" API, aligning it with the API design described here and agreed to here and here.

In particular this changes the v2 API from:

```
    POST /collections/collName {"add-replica-property" : {...}}
```

to...

```
PUT /collections/<coll>/shards/<shard>/replicas/<replica>/properties/<prop> {"value": <val>}
```

This is a breaking change for users of the v2 API, but one that is allowed due to v2's "experimental" designation.

# Solution

This PR modifies add-replica-prop in the way described, but it also refactors the API class to use the newly added JAX-RS framework in the process.  This framework-switchover, while conceptually separate from the API realignment in this case, makes sense as both tasks were coming in the short term and touch identical chunks of code.

As a result, this PR not only changes the cosmetics of the add-replica-prop endpoint, it also adds it to Solr's growing OpenAPI spec.

As suggested in the past on similar refactors, I've also moved the logic underlying the add-replica-prop functionality (previously hosted in `CollectionsHandler`) over to the v2 `AddReplicaPropertyAPI` class.  The v1 codepath in `CollectionsHandler` now instantiates an AddReplicaPropertyAPI to do its job. 

# Tests

New unit tests for `AddReplicaPropertyAPI`, as well as all the existing v1 and v2 tests for the add-replica-prop functionality continuing to pass.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ ] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

gerlowskija committed:
SOLR-13692: Convert ADDREPLICAPROP API to JAX-RS

This commit updates ADDREPLICAPROP to JAX-RS, changing the endpoint
cosmetics to match our more REST-ful design in the process.  The v2
endpoint appears to work, but the following changes still need made:

  - update existing and add new tests
  - v1 codepath needs updated to use Jersey resource
  - OpenAPI annotations need added.

gerlowskija committed:
SOLR-16392: Refactor+update v2 ADDREPLICAPROP API

This commit modifies the existing AddReplicaPropertyAPI definition to be
more REST-ful, and use the new JAX-RS framework. Specifically, this
commit changes the API from:

POST /collections/collName
    {"add-replica-property" : {...}}

to

PUT /collections/<coll>/shards/<shard>/replicas/<replica>/properties/<prop>
    {"value": <val>}

This is a breaking change for users of the v2 API, but one that is
allowed because of v2's "experimental" designation.

In support of this, this commit also refactors AddReplicaPropertyAPI to
use the new JAX-RS framework.  It also introduces the new class,
"InjectionFactories" to allow Jersey to inject common "context"
values/objects as ctor args for resource classes such as
AddReplicaPropertyAPI.

gerlowskija committed:
Add AddReplicaPropertyAPI unit tests

gerlowskija commented:
This PR is blocked on merging apache/solr PR #1035, but is otherwise ready for review!

epugh commented on solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java:
```diff
@@ -298,37 +303,59 @@ public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throw
-      }
-      invokeAction(req, rsp, cores, action, operation);
-    } else {
+    if (a == null) {
```
I don't know how much celan up you want....   but `a` as a variable?   Makes me think there must be a `b` somewhere around ;-).   `actionParam`?  `action`?

gerlowskija replied:
Agreed.  IMO it comes back to your first comment though: "I don't know how much clean up you want".

Since we're aiming to make similar changes across all of Solr's 200+ APIs, I'm leery of straying into any sort of cleanup that's not strictly necessary for our API goals - even though I totally agree that it's absolutely needed in general.

Even IDE refactors like variable renaming have a "cost", as they dirty diffs and make it marginally harder for folks to review.

epugh replied:
Yeah...   You are right...  sigh...   Good for you for staying on point and not letting me distract you!   (This is why my PR's tend to grow over time :-(  ).    

Any thoughts on how we could track these for future fixing?   I worry about creating a million JIRA's...    Notes in the spreadsheet "come back and check this"?

epugh commented on solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java:
```diff
@@ -298,37 +303,59 @@ public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throw
     }
-    rsp.setHttpCaching(false);
-  }
+    CollectionAction action = CollectionAction.get(a);
```
ah...   Now we are in a more comfortable world...  an Action ;-)

epugh commented on solr/core/src/java/org/apache/solr/handler/admin/CollectionsHandler.java:
```diff
@@ -386,8 +413,13 @@ public SolrResponse submitCollectionApiCommand(ZkNodeProps m, CollectionAction a
+      CoreContainer coreContainer,
+      Optional<DistributedCollectionConfigSetCommandRunner>
+          distributedCollectionConfigSetCommandRunner,
+      ZkNodeProps m,
```
`m`?  Maybe `zkNodeProps`?

gerlowskija replied:
Agreed, but see comment [here](https://github.com/gerlowskija/solr/pull/1#discussion_r981144239).

gerlowskija committed:
Remove unused payload class

gerlowskija committed:
Remove unused CollectionOperation definition

gerlowskija committed:
Merge branch 'main' into SOLR-16392-addreplicaprop-conversion-and-cosmetics

gerlowskija committed:
Add CHANGES.txt entry

gerlowskija committed:
Make tidy happy

gerlowskija committed:
SOLR-16392: Support REST-ful path lookup in V2HttpCall (#1035)

Prior to this commit, V2HttpCall uses a rather crude heuristic to
determine whether a given API is a "container-level" or "core-level"
API.  Namely, any path that a core/collection name can be parsed out of
is a "core" API that should live in the core-level PluginBag.
Conversely, anything else is a container-level API that should live in
the container-level PluginBag.

This happened to work for V2 because we initially chose our API paths
with this heuristic in mind.  But the result is an API that's neither intuitive
nor user-friendly.

This PR takes a sort of "guess and check" approach, where we try
serving ambiguous requests from multiple Jersey apps in series, stopping
at the first one that doesn't spit out a quick NotFoundException.

gerlowskija committed:
Merge branch 'main' into SOLR-16392-addreplicaprop-conversion-and-cosmetics

gerlowskija commented:
This PR was originally stacked on top of another [PR](https://github.com/apache/solr/pull/1035) that has since been merged into upstream's 'main'.  But I can't find a way to convert this PR over to apache/solr's 'main' branch - just the main branch of my own fork.

So rather than fight github on this, I'm going to close out this PR and create another one.  Sorry to all for the loss of history 😬
